packages/blocks/src/api/factory.js

```diff
@@ -50,6 +50,10 @@ export function createBlock( name, blockAttributes = {}, innerBlocks = [] ) {
 			result[ key ] = schema.default;
 		}
 
+		if ( schema.source === 'html' && typeof result[ key ] !== 'string' ) {
+			result[ key ] = '';
```

Comment:
Isn't this change a little bit risky? e.g: some block may be comparing the attribute with undefining to check if the value does not exist and that may stop working now.
Also isn't there a risk a block may want to differentiate an HTML attribute being undefined (e.g: the query element was not found ) with an HTML attribute being empty (element exists but its content is empty )?
In this case, I would be inclined to just add a default value of the empty string in the merge function to reduce the risks.

Reply:
> (e.g: the query element was not found )

We currently already pass `''` if the query is not found.

Alternative is to add `default: ''` everywhere.

Reply:
And do you have a use case for the `undefined`? Note that `! value` will still work.

Reply:
If we already pass '' if the query was not found I think everything seems fine. I was not aware of that.

Reply:
https://github.com/WordPress/gutenberg/blob/d374e0222332fc519e12ec43a15e31aea3114ca2/packages/blocks/src/api/matchers.js#L21

Reply:
@aduth @youknowriad - can you do sanity check later. I'm fine with merging it as is to get 4.0 out.

Reply:
@jorgefilipecosta Could you create a new issue with a use case fro block authors for `undefined`?

Reply:
I'd prefer `default` everywhere or accounting for the `undefined` value in merges much more than to include ad hoc corrections in the factory. 

It's also conflicting messaging when we went out of our way in #10338 to explicitly remove tolerances. If we'd kept the tolerances, I'd rather see something where we normalize `undefined` of `type: 'string'` to an empty string, but again still nothing targeting one specific `source`.

What's the issue with providing the `default`, aside from tedium / verbosity? I see it as no different (and thus extremely confusing in its consistency) from how we don't provide empty arrays for array types, zero's for number types, or an empty string for any other string types (what's the difference here between why we need an empty string here for assumed HTML stringiness and `source: 'attribute'`?).

At worst, I'd think we should codify this as a `source`-level `default` mapping.

Reply:
I find it a bit weird that plugin authors will have to provide the default value, and may forget about it. Otherwise we have to set the default in merge functions etc., but since even we forget about it, I think block authors might also forget about it and cause these preventable bugs.

Reply:
That's all well and good as a goal, but targeting a resolution for something so specific is going to have negative consequences (i.e. bugs) elsewhere in setting people up for mixed expectations for what is and isn't normalized on their behalf.

Reply:
Okay, then let's change it. I'm not sure what the right solution is here tbh.